columnflow/tasks/ml.py

```diff
@@ -445,9 +603,32 @@ class MLEvaluation(
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
+        # cache for producer inst
+        self._preparation_producer_inst = None
+
         # set the sandbox
         self.sandbox = self.ml_model_inst.sandbox(self)
 
+    @property
+    def preparation_producer_inst(self):
+        producer = None
+        if self.ml_model_inst.preparation_producer_in_ml_evaluation:
+            producer = self.ml_model_inst.preparation_producer(self.config_inst)
+        if producer and self._preparation_producer_inst is None:
```

Comment:
Same as above, restructure such that nothing is executed twice when already cached.